File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        if(s.length() % 2 != 0) return false;
        
        Stack<Character> sc = new Stack<>();
        for(char c : s.toCharArray()) {
            if(c == '(' || c == '{' || c == '[') {
                sc.push(c);
            }
            else if(c == ')' && !sc.isEmpty() && sc.peek() == '(') {
                sc.pop();
            }
            else if(c == '}' && !sc.isEmpty() && sc.peek() == '{') {
                sc.pop();
            }
            else if(c == ']' && !sc.isEmpty() && sc.peek() == '[') {
                sc.pop();
            }
            else {
                return false;
            }
        }
        return sc.isEmpty();
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        st = []
        parenthesis = {')' : '(', ']' : '[', '}' : '{'}

        for char in s:
            if char in parenthesis:
                if st and st[-1] == parenthesis[char]:
                    st.pop()
                else:
                    return False
            else:
                st.append(char)

        return len(st) == 0